lib/types/array.js

```diff
@@ -661,14 +661,18 @@ internals.fillMissedErrors = function (schema, errors, requireds, value, state,
 };
 
 
-internals.fillOrderedErrors = function (schema, errors, ordereds, value, state, prefs) {
+internals.checkRemainingOrdereds = function (schema, errors, ordereds, value, state, prefs) {
 
     const requiredOrdereds = [];
 
     for (const ordered of ordereds) {
         if (ordered._flags.presence === 'required') {
             requiredOrdereds.push(ordered);
         }
+
+        if (ordered._flags.hasOwnProperty('default')) {
```

Comment:
This doesn't check for dynamic schemas, such as links or schemas with when conditions. It also doesn't resolve references.

Reply:
I'd suggest calling `$_validate()` with `undefined` to make sure the full validation cycle is invoked. Also, it would probably make more sense to "fill" the required errors first before filling default values.

Reply:
Hi @brianle1301 Thanks for making time reviewing this. 
> This doesn't check for dynamic schemas, such as links or schemas with when conditions. It also doesn't resolve references.
Edit: Also doesn't work for objects marked with the deep default symbol.

My mistake, i didn't think about using a dynamic schema as a default value. I will change it with
```JS
        const res = ordered.$_validate(undefined,state,prefs);
        if (!res.errors){
            orderedsWithDefault.push(res.value)
        }
```

and then maybe filling the defaults after filling the required error this way: 
```JS
  if (orderedsWithDefault.length) {
        value.push(...orderedsWithDefault);
    }
```

> Instead of automatically filling default values, you need to explicitly call default() on the array schema to trigger this process

I didn't get what you mean by ` explicitly call default() on the array schema`, can you elaborate more?

Reply:
@iifawzi The check `if (!res.errors)` is probably not needed, because the only way in which a schema will throw on `undefined` is if it's marked with `required()` (I will need to verify this again to make sure). And because you are filling required errors first, it is guaranteed that the ordered schemas will not have `required()` marked on them (otherwise, the Joi would bail and return the required errors).

I also think we need to consider cases where filling defaults would generate sparsed arrays and let users customise how they want their final arrays to look. For example, this schema 
```
Joi.array().ordered(Joi.string().default('test'), Joi.string(), Joi.number().default(5))
```
would generate `['test', undefined, 5]`, which sometimes isn't desirable. 

To answer your second question, Joi generates deep default on object schemas if marked with `default()` (without any arguments). I would like to see something similar for arrays, that is, required values are only filled when `default()` is called on them. However, it would probably be better if you get everything working first because it's reasonably easy to refactor to make the above possible. 

Make sure you also cc Eran early for his reviews so not too much time is wasted. Kinda risky to work on I'd say is quite "complex" 😄

Reply:
> The check if (!res.errors) is probably not needed, because the only way in which a schema will throw on undefined is if it's marked with required() (I will need to verify this again to make sure). And because you are filling required errors first, it is guaranteed that the ordered schemas will not have required() marked on them (otherwise, Joi would bail and return the required errors).

I'm waiting for your confirmation, I checked the `validate` function, and it seems like it will not throw an error yes if the value is `undefined` and the schema is not marked with `required` but waiting for your confirmation for this, i wouldn't be surprised if missed any. 

> I also think we need to consider cases where filling defaults would generate sparsed arrays and let users customise how they want their final arrays to look. For example, this schema would generate ['test', undefined, 5], which sometimes isn't desirable.
```JS
Joi.array().ordered(Joi.string().default('test'), Joi.string(), Joi.number().default(5))
```

- You're right, if i added a check that the value is not `undefined` would be enough i think, is there's any other option? if it's not marked as `required` or has a `default` value so it shouldn't be added at all (logically), as if it's not exists. And that's what really happens in other cases when the value is not inserted for a field with no `required` constraint (or this at least what i noticed when using Joi)

> To answer your second question, Joi generates deep default on object schemas if marked with default() (without any arguments). I would like to see something similar for arrays, that is, required values are only filled when default() is called on them. However, it would probably be better if you get everything working first because it's reasonably easy to refactor to make the above possible.
Make sure you also cc Eran early for his reviews so not too much time is wasted. Kinda risky to work on I'd say is quite "complex" 😄

Thanks for the clarification, you're right. I walk though getting everything working first then looking into it (and to be honest i didn't fully get it until now, but will look into it later after getting done with the essential part). 

@hueniverse Hi Eran, can you take look on this if you got a time? would it be hard to cover all cases? do you prefer not working on it? 

- I committed the earlier changes which @brianle1301 mentioned and i added the `undefined` check for the moment.